opnode/node/driver.go

```diff
@@ -0,0 +1,87 @@
+package node
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/trie"
+	"github.com/holiman/uint256"
+)
+
+type InputItem struct {
+	Block    *types.Block
+	Receipts []*types.Receipt
+}
+
+func (item *InputItem) Epoch() uint64 {
+	return item.Block.NumberU64()
+}
+
+// sanity check that the receipts are consistent with the block data
+func (item *InputItem) CheckReceipts() bool {
+	hasher := trie.NewStackTrie(nil)
+	computed := types.DeriveSha(types.Receipts(item.Receipts), hasher)
```

Comment:
the `types.Receipts(...)` performs an array copy, right?

Reply:
No it does not. Go distinguishes between `slice` and `array`. This is just a slice: the type contains the element type, the memory is a pointer to the start, the slice length (element count, not memory size), and capacity (element count of underlying array that backs the slice). The memory of the slice-header is copied, but is only 24 bytes. The type is cast and doesn't cost us anything. This cast works because the element type is the same, and it's only the slice type that changes.